Drivers/CAN/can.c

```diff
@@ -0,0 +1,289 @@
+            CAN_message.data[i] = 0; // copies data to message, padding with 0s if length isn't 8
+        }
+    }
+    switch(bus)
+    {
+        case CAN_line_1:
+            CAN_message.CAN_handler = CAN1;
+            break;
+        case CAN_line_2:
+            CAN_message.CAN_handler = CAN2;
+            break;
+        case CAN_line_3:
+            CAN_message.CAN_handler = CAN3;
+            break;
+    }
+
+    CAN_TxHeaderTypeDef* tx_header = malloc(sizeof(CAN_TxHeaderTypeDef));
+    tx_header->StdId = id;
+    tx_header->ExtId = CAN_EXT_ID;
+    tx_header->IDE = CAN_ID_STD;
+    tx_header->RTR = CAN_RTR_data;
+    tx_header->DLC = len;
+    tx_header->TransmitGlobalTime = DISABLE;
+    CAN.CAN_tx_header = tx_header;
+
+    CAN_RxHeaderTypeDef* rx_header = malloc(sizeof(CAN_RxHeaderTypeDef));
+    rx_header->StdId = id;
+    rx_header->ExtId = CAN_EXT_ID;
+    rx_header->IDE = CAN_ID_STD;
+    rx_header->RTR = CAN_RTR_data;
+    rx_header->DLC = len;
+    rx_header->Timestamp = 0;
+    rx_header->FilterMatchIndex = 0;
+    CAN.CAN_rx_header = rx_header;
+    return CAN;
+}
+
+int sendMessageCAN1(uint32_t id, uint8_t len, const uint8_t *data)
+{
+    CAN_t message = CAN_setup(id, len, data, CAN_line_1, CAN1);
+    uint32_t free_spots = HAL_CAN_GetTxMailboxesFreeLevel(message.CAN_handler);
+    if(free_spots != 0)
+    {
+        HAL_StatusTypeDef bruh = HAL_CAN_AddTxMessage(message.CAN_handler, message.CAN_transmit_header, message.data, CAN_TX_MAILBOX0);
+        if(bruh == HAL_OK)
+        {
+            return SUCCESS;
+        }
+        else
+        {
+            return FAIL;
+        }
+    }
+    else
+    {
+        return FAIL;
+    }
+}
+
+int sendMessageCAN2(uint32_t id, uint8_t len, const uint8_t *data)
+{
+    CAN_t message = CAN_setup(id, len, data, CAN_line_2, CAN2);
+    uint32_t free_spots = HAL_CAN_GetTxMailboxesFreeLevel(message.CAN_handler);
+    if(free_spots != 0)
+    {
+        HAL_StatusTypeDef bruh = HAL_CAN_AddTxMessage(message.CAN_handler, message.CAN_transmit_header, message.data, CAN_TX_MAILBOX1);
+        if(bruh == HAL_OK)
+        {
+            return SUCCESS;
+        }
+        else
+        {
+            return FAIL;
+        }
+    }
+    else
+    {
+        return FAIL;
+    }
+}
+
+int sendMessageCAN3(uint32_t id, uint8_t len, const uint8_t *data)
+{
+    CAN_t message = CAN_setup(id, len, data, CAN_line_3, CAN3);
+    uint32_t free_spots = HAL_CAN_GetTxMailboxesFreeLevel(message.CAN_handler);
+    if(free_spots != 0)
+    {
+        HAL_StatusTypeDef bruh = HAL_CAN_AddTxMessage(message.CAN_handler, message.CAN_transmit_header, message.data, CAN_TX_MAILBOX2);
+        if(bruh == HAL_OK)
+        {
+            return SUCCESS;
+        }
+        else
+        {
+            return FAIL;
+        }
+    }
+    else
+    {
+        return FAIL;
+    }
+}
+
+// TODO: Make a big switch statement of all the handler functions
+uint8_t[8] incoming_callback_CAN(uint8_t bus, uint32_t msg_id)
```

Comment:
I still don't think we want any callbacks on this level of driver. If there are any, there should only be one main callback function that literally just returns and does nothing that the user can replace later. We shouldn't implement any specific message handlers. Remember that this might be used across multiple STM projects.

Reply:
am i stupid or does this clalback not have a callback in it